GPIO Hogs: add support for masking the output level

keith-zephyr opened the pull request:
Allow applications to prevent the GPIO hogs driver from changing the physical state of a pin during GPIO configuration. 

One potential use case is an application that initiates a software jump between Zephyr images.  The first image that runs from the power on reset configures all GPIO hogs normally, but the second image doesn't want to change the physical state of any output pins, which might have side effects on the operation of the board.

keith-zephyr committed:
test: gpio_hogs: Test active high signals

Update the GPIO hogs test to verify output GPIOs configured for both
active high and active low configuration.

Signed-off-by: Keith Short <keithshort@google.com>

keith-zephyr commented:
Looking for some feedback on this approach.  I have a downstream application that needs this capability.

henrikbrixandersen reviewed (changes requested):
I am not sure I fully understand this. Why not just exclude the GPIO hogs from the DTS of the chain-loaded Zephyr application?

keith-zephyr commented:
> I am not sure I fully understand this. Why not just exclude the GPIO hogs from the DTS of the chain-loaded Zephyr application?

For our application, we have two copies of the firmware for security.
* Image A (considered read-only, never updated)
* Image B (can be updated in the field)

Normally, Image A boots from the power on reset, validates Image B and then jumps to Image B.

When we do a firmware update, Image B jumps back to Image A first, so we can then program a new Image B.

So in this case Image A, is required to configure GPIOs from power on reset, but it needs to skip this step if it is jumped to from Image B.

henrikbrixandersen commented:
> For our application, we have two copies of the firmware for security.
> 
> * Image A (considered read-only, never updated)
> * Image B (can be updated in the field)
> 
> Normally, Image A boots from the power on reset, validates Image B and then jumps to Image B.
> 
> When we do a firmware update, Image B jumps back to Image A first, so we can then program a new Image B.
> 
> So in this case Image A, is required to configure GPIOs from power on reset, but it needs to skip this step if it is jumped to from Image B.

Could a solution be to leave GPIO hogs configuration disabled for Image A and enabled for Image B?

I must admit that I believe this is too much of a corner case to be addressed in upstream Zephyr due to the added complexity. What if someone requires a specific UART driver to only be initialised if coming out of a watchdog reset? This approach won't scale. If the above suggestion about only using GPIO hogs in one of the images cannot solve the issue, I would suggest using a board/application-specific solution not involving the GPIO hogs driver.

keith-zephyr commented:
> Could a solution be to leave GPIO hogs configuration disabled for Image A and enabled for Image B?
> 
> I must admit that I believe this is too much of a corner case to be addressed in upstream Zephyr due to the added complexity. What if someone requires a specific UART driver to only be initialised if coming out of a watchdog reset? This approach won't scale. If the above suggestion about only using GPIO hogs in one of the images cannot solve the issue, I would suggest using a board/application-specific solution not involving the GPIO hogs driver.

If Image A fails to validate Image B, then Image A must still configure the GPIOs. 

What do think of a Kconfig option that makes `gpio_hogs_init()` public?  When this new option is enabled, the current `SYS_INIT()` call is compiled out and it would be up to the application to call or skip `gpio_hogs_init()`.

This may not satisfy all my application needs, but I might be able to work within that framework.  I have a downstream solution now, but I'd prefer to utilize upstream drivers capabilities when I can.

fabiobaltieri commented on drivers/gpio/Kconfig:
```diff
@@ -59,6 +59,22 @@ config GPIO_HOGS_INIT_PRIORITY
 
+config GPIO_HOGS_MASK_OUTPUT
+	bool "GPIO hogs mask output"
+	default 0
```
default n?

cfriedt replied:
The default should just be left out. It defaults to n by default 😄

fabiobaltieri commented on drivers/gpio/gpio_hogs.c:
```diff
@@ -90,16 +89,33 @@ static const struct gpio_hogs gpio_hogs[] = {
+	uint32_t reset_cause;
+
+	if ((hwinfo_get_reset_cause(&reset_cause) == 0) &&
+	    (reset_cause & CONFIG_GPIO_HOGS_RESET_CAUSE_MASK)) {
```
Tying this to a reset cause mask feels a bit oddly specific. The idea of having the application code `gpio_hogs_init` may be good though. Maybe you could have an option for auto initializing the hog in Kconfig (similar to `CONFIG_USB_DEVICE_INITIALIZE_AT_BOOT`), and then an argument to `gpio_hogs_init` to skip level initialization? Then the application can initialize manually and use any logic for the argument.

keith-zephyr replied:
Good suggestion, let me try this out.

keith-zephyr committed:
gpio: hogs: Add app driven initialization option

Add the CONFIG_GPIO_HOGS_INITIALIZE_BY_APPLICATION option so that
applications can make a run-time decision whether to initalize the GPIO
hogs or to mask certain GPIO flags during initialization.

Signed-off-by: Keith Short <keithshort@google.com>

keith-zephyr committed:
gpio: hogs: Add support for storing line-names

Change the line-name property to line-names and optionally store the
lines-names into the GPIO hogs structure.

If the line-names property is omitted, the name defaults to full node
name appended with "_idx".

Signed-off-by: Keith Short <keithshort@google.com>

keith-zephyr committed:
gpio: hogs: Add shell command

Add a shell command for getting and setting the GPIO pin state for GPIOs
mapped to GPIO hogs.

Signed-off-by: Keith Short <keithshort@google.com>

keith-zephyr committed:
gpio: hogs: Bypass config for pins without in/out

Skip configuring any GPIO hog that omits all of the input, output-high,
and output-low properties.

This is useful in conjunction with the GPIO hogs shell command to make a
GPIO available to the shell commands, but the application or driver is
still responsible for configuring the pin.

Signed-off-by: Keith Short <keithshort@google.com>
